courses/codepeer/010_codepeer.rst

```diff
@@ -0,0 +1,2666 @@
++ Any **combination** desktop/continuous/nightly run
++ **Per-project** software customization
++ **Compare** local changes with master
++ Multiple teams **reviewing** multiple subsystems
++ Use :toolname:`CodePeer` to generate a **security report**
+
+----------------------------------------------
+Analyzing Code Locally Prior To Commit (1/2)
+----------------------------------------------
+
++ Each **developer** as a single user, on a **desktop** machine
++ After compilation, before testing.
++ Solution #1: File by File analysis
+
+  + Use :toolname:`GNAT Studio` menu
+  + :menu:`CodePeer` :math:`\rightarrow` :menu:`Analyze File`
+  + On the files that were **modified**
+  + Fastest, incremental
+
++ Solution #2
+
+  + Run :command:`codepeer -level 1/2 -baseline`
+  + Local **baseline** database used for comparison
+  + Look at **added** messages only
+  + More exhaustive
+  + Uses past reviews (less false positives)
+
+----------------------------------------------
+Analyzing Code Locally Prior To Commit (2/2)
+----------------------------------------------
+
++ If duration or number of messages is not good :math:`\rightarrow` refine the settings
++ For each new message:
+
+   + If a real issue is found :math:`\rightarrow` Fix the code
+   + If it is a false positive :math:`\rightarrow` Justify it with :ada:`pragma Annotate`
+
+--------------
+Nightly Runs
+--------------
+
++ :toolname:`CodePeer` run daily on a dedicated server
+
+    + With large resources
+    + Exhaustive level (2 :math:`\rightarrow` 4)
+
++ Typically run nightly
+
+    + Takes into account commits of the day
+    + Provides results to users the next morning
+
++ Allows users to analyze and justify messages **manually**
+
+    + Via the **web** interface
+    + From :toolname:`GNAT Studio` by accessing the **database** remotely
+
++ At release, results can be committed under CM for **traceability** purposes
+
+-----------------
+Continuous Runs
+-----------------
+
++ :toolname:`CodePeer` is run on a dedicated server
+
+    + With large resources
+    + Fast level (0 or 1)
+
++ No need to be exhaustive
+
+    + Focus on **differences** from previous run
+
++ Continuous runs triggerred on repository events
++ Summary is sent to developers
+
+    + Email
+    + Web interface
+      :command:`codepeer -Pprj -output-msg -only -show-added | grep "[added]"`
+
++ Developers then *fix the code*, or *justify the relevant messages*
+
+  + via :ada:`pragma Annotate` in source code or via web interface.
+  + or wait for the next nightly run to post a manual analysis via the HTML Output.
+
+------------------------------
+Combined Desktop/Nightly Run
+------------------------------
+
++ **Fast** analysis of code changes done at each **developer's desk**
++ A longer and **more exhaustive** analysis is performed nightly
++ The developer can re-use the **nightly** database as a baseline for analysis
++ Database reviews **should** be stored in the **nightly** (*gold*) database
```

Comment:
What does "gold" mean? (We never describe/explain)

Reply:
Removing it in most cases